src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnectionOptions.cs

```diff
@@ -66,6 +76,29 @@ public HttpConnectionOptions()
             set => _headers = value ?? throw new ArgumentNullException(nameof(value));
         }
 
+        /// <summary>
+        /// Gets or sets the maximum buffer size for data read by the application before backpressure is applied.
+        /// </summary>
+        /// <remarks>
+        /// The default value is 1MB.
+        /// </remarks>
+        public long TransportMaxBufferSize { get; set; }
+
+        /// <summary>
+        /// Gets or sets the maximum buffer size for data written by the application before backpressure is applied.
+        /// </summary>
+        /// <remarks>
+        /// The default value is 1MB.
+        /// </remarks>
+        public long ApplicationMaxBufferSize { get; set; }
+
+        // We initialize these lazily based on the state of the options specified here.
+        // Though these are mutable it's extremely rare that they would be mutated past the
+        // call to initialize the routerware.
+        internal PipeOptions TransportPipeOptions => _transportPipeOptions ??= new PipeOptions(pauseWriterThreshold: TransportMaxBufferSize, resumeWriterThreshold: TransportMaxBufferSize / 2, readerScheduler: PipeScheduler.ThreadPool, useSynchronizationContext: false);
```

Comment:
Should we disallow `-1` for these sizes? I don't think this is unique to SignalR, but after discussing it in API review, I realize this would lead to weird behavior where you'd get System.IO.Pipeline's default `pauseWriterThreshold` (64k), but not the default resumeWriterThreshold (0 instead of 32k) meaning *all* the data would need to be consumed before backpressure is relieved.

Reply:
Oh I thought -1 was unlimited, but it's defaults. It might make sense for the API to not allow -1 then.